mlir/lib/Driver/CompilerDriver.cpp

```diff
@@ -274,79 +286,108 @@ LogicalResult runEnzymePasses(const CompilerOptions &options,
     if (options.keepIntermediate) {
         llvm::raw_string_ostream rawStringOstream{outputs["Enzyme"]};
         llvmModule->print(rawStringOstream, nullptr);
-        const std::string &outFile = fs::path("2_Enzyme.ll");
-        if (failed(dumpToFile(options, outFile, outputs["Enzyme"]))) {
-            return failure();
-        }
+        auto outFile = output.nextPipelineDumpFilename("Enzyme", ".ll");
+        dumpToFile(options, outFile, outputs["Enzyme"]);
     }
 
     return success();
 }
 
 LogicalResult runLowering(const CompilerOptions &options, MLIRContext *ctx, ModuleOp moduleOp,
-                          CompilerOutput::PipelineOutputs &outputs)
+                          CompilerOutput &output)
+
 {
+    auto &outputs = output.pipelineOutputs;
     auto pm = PassManager::on<ModuleOp>(ctx, PassManager::Nesting::Implicit);
 
+    // Maps a pass to one or more pipelines ended by this pass
+    // Maps a pass to its owning pipeline
+    // Lists pipelines not terminated by any passes
     std::unordered_map<const Pass *, std::list<Pipeline::Name>> pipelineTailMarkers;
+    std::unordered_map<const Pass *, Pipeline::Name> passPipelineNames;
+    std::vector<Pipeline::Name> dunglingPipelines;
+
+    // Fill all the pipe-to-pipeline mappings
     for (const auto &pipeline : options.pipelinesCfg) {
+        size_t existingPasses = pm.size();
         if (failed(parsePassPipeline(joinPasses(pipeline.passes), pm, options.diagnosticStream))) {
             return failure();
         }
-        PassManager::pass_iterator p = pm.end();
-        const Pass *lastPass = &(*(p - 1));
-        pipelineTailMarkers[lastPass].push_back(pipeline.name);
+        if (existingPasses == pm.size()) {
+            dunglingPipelines.push_back(pipeline.name);
+        }
+        else {
+            const Pass *pass = nullptr;
+            for (size_t pn = existingPasses; pn < pm.size(); pn++) {
+                pass = &(*(pm.begin() + pn));
+                passPipelineNames[pass] = pipeline.name;
+            }
+            assert(pass != nullptr);
+            for (auto pn : dunglingPipelines) {
+                pipelineTailMarkers[pass].push_back(pn);
+            }
+            dunglingPipelines.clear();
+            pipelineTailMarkers[pass].push_back(pipeline.name);
+        }
     }
 
-    size_t pipelineIdx = 0;
     if (options.keepIntermediate) {
-
+        std::string tmp;
         {
-            std::string tmp;
-            {
-                llvm::raw_string_ostream s{tmp};
-                s << moduleOp;
-            }
-            const std::string &outFile =
-                fs::path(options.moduleName.str()).replace_extension(".mlir");
-            if (failed(dumpToFile(options, outFile, tmp))) {
-                return failure();
-            }
+            llvm::raw_string_ostream s{tmp};
+            s << moduleOp;
         }
```

Comment:
Are you also suggesting to remove the dumpFile call? I think we should limit the lifetime of `s` because we need it to flush the stream before we pass it to the dump function.

Reply:
I see. I made a mistake, I deleted `dumpToFile`. Sorry about that!

I am looking at the [documentation](https://llvm.org/doxygen/classllvm_1_1raw__string__ostream.html) for `llvm::raw_string_ostream` and it says the following:

> This is a simple adaptor class. This class does not encounter output errors. [raw_string_ostream](https://llvm.org/doxygen/classllvm_1_1raw__string__ostream.html) operates without a buffer, delegating all memory management to the std::string. Thus the std::string is always up-to-date, may be used directly and there is no need to call [flush()](https://llvm.org/doxygen/classllvm_1_1raw__ostream.html#a520bdf57dfe3e73abb53d482893f0a27).

Reply:
> Are you also suggesting to remove the dumpToFile call? 

Looking back, I don't think I am suggesting to remove `dumpToFile`. `dumpToFile` is still below. All I'm suggesting is to remove the braces that limit the scope of `s`

Reply:
Regarding dumpToFile - Yes, my bad, sorry. That was my own change.

Thanks for checking the docs, the scope is indeed not needed if it doesn't use buffers.  I will remove it soon.

Reply:
Done